### ISSUE TITLE
(Discussion) Equinox support

### ISSUE THREAD

**Issue by clementpoiret** [enpasos/jax2onnx#39]
Hi,

jax2onnx already supports raw Jax and Flax.
Equinox being also quite popular, do you have any plans about adding equinox to the stack of supported frameworks?

Thanks, and have a nice day!

**Comment by enpasos** [comment]
Sounds great! I’ve pushed an initial example to the `equinox` branch.

You can find the plugin here: `jax2onnx/plugins/equinox/eqx/linear.py`

And an example setup here: `jax2onnx/plugins/examples/eqx/simple_linear.py`
This includes a minimal `equinox.nn.Linear` model, and you can build on it to test models with more components.

To hook it into the test suite, run once:

```bash
poetry run python scripts/generate_tests.py
```

Then execute tests normally via:

```bash
poetry run pytest
```

You can regenerate the README with:

```bash
poetry run python scripts/generate_readme.py
```

Note: commits are gated by pre-commit checks (`poetry run pre-commit run --all-files`), mainly for the main branch. Feel free to disable them during development.

Good luck – the equinox branch is all yours! Let me know if you need anything along the way.

**Comment by clementpoiret** [comment]
Okay so I took some time to check your example for the linear layer. To start with something easy [I added the `eqx.nn.Identity`](https://github.com/enpasos/jax2onnx/pull/40).
When I check the produced onnx graphs on Netron, I see that I have the correct onnx identity node. But when I look at the linear layer, I see the whole lowered graph, although the same onnx files for Flax are a simple `Gemm` node. Do you know why?
Thanks!

**Comment by enpasos** [comment]
Sorry for leaving you with a not-quite-working starting point — the `batching_rule` wasn’t implemented correctly, and there wasn’t a test in place to catch it. It should be working now — good luck!

**Comment by clementpoiret** [comment]
Thanks for your update 👌 
I just accepted the invite

**Comment by clementpoiret** [comment]
I can confirm I can push on equinox, thanks!

**Comment by clementpoiret** [comment]
Quick question: I am trying to convert the dropout layer. Both on the flax and my [equinox version](https://github.com/enpasos/jax2onnx/blob/equinox/jax2onnx/plugins/equinox/eqx/dropout.py), Netron is showing `training_mode=true`, even when `deterministic=True` (flax), or `inference=True` (equinox), but tests on training mode are all passing. Do you know why?

Also (and final question, sorry to bother you), I have troubles passing a dynamic inference to a Not gate then as the 3rd input of Dropout (as you have done for flax). When I look at my generated graph, `inference` is never linked to the dropout node. Do you know what I am doing wrong? Thanks!

**Comment by enpasos** [comment]
I’m looking into it, starting with `testcase=dropout_init_params`. The obvious issue is that the flag in the ONNX model is definitely wrong. But with `ratio=0.5`, the test really *should* fail based on my current understanding. It seems like my assumptions about how the ONNX Runtime handles this operator might be off. I’ll need to dig deeper.

**Comment by enpasos** [comment]
I dug deeper—and the outcome is surprising: it appears to be a **bug in Netron**.

When I open `dropout_init_params.onnx` in Netron, I see this:

![Image](https://github.com/user-attachments/assets/0e143c89-2f02-4cd9-9a1a-706b018af6b9)

However, when I run a minimal ONNX inspection script:

```python
import numpy as np
import onnx
from onnx import numpy_helper

model_path = "dropout_init_params.onnx"
onnx_model = onnx.load(model_path)
graph = onnx_model.graph

for init in graph.initializer:
    arr = numpy_helper.to_array(init)
    if arr.ndim == 0 and arr.dtype == np.bool_:
        print(f"  {init.name:<20}: {arr}  (dtype={arr.dtype})")
```

…the output clearly states:

```
  const_1             : False  (dtype=bool)
```

Still skeptical, I opened a hex editor and inspected the raw protobuf bytes manually. The structure matches the Protocol Buffers spec for a `TensorProto` with a scalar `BOOL` value of `False` (`0x00`), confirming that the Python inspection is accurate.

I'll go ahead and open a ticket on the Netron repository to clarify this discrepancy. Whether it's a Netron bug or a subtle quirk in interpretation, I'm confident Lutz Roeder will help shed light on it.

**Comment by enpasos** [comment]
Here's the link to the Netron issue I opened: [https://github.com/lutzroeder/netron/issues/1470](https://github.com/lutzroeder/netron/issues/1470)

**Comment by enpasos** [comment]
> For the dynamic inference part—mind sharing the code you’re hanging with? Then I can take a closer look and help untangle it.

I just spotted the code on the `equinox` branch — I’ll have a look there!

**Comment by enpasos** [comment]
I’ve pushed a fix for the inference connection issue.

I’m fully aware that the utility layer used across plugins has major room for improvement. If it doesn’t feel intuitive, that’s a valid impression — and something we should definitely work on.

---

### 🔧 Summary of Changes

#### 1. **Proper Wiring of Dynamic `inference` Input**

The main issue was that the inverted `inference` flag (via `Not`) wasn’t being correctly connected as the third input to the Dropout node. The previous logic attempted to rename the input but didn’t reliably link the `Not` output or register shape/type metadata.

**Fix:**

* Renamed the JAX `Var` to `"inference"` using `s.change_var_name`.
* Wired the `Not` node correctly.
* Registered the shape and type for the new `training_mode` tensor using `s.add_shape_info`.

#### 2. **Modernized JAX API Usage**

The code previously referenced `core.Var`, which is deprecated. This caused issues in newer JAX versions.

**Fix:**

* Switched to `Var` from `jax.extend.core`.

#### 3. **Improved Test for Dynamic Inference**

The test case now explicitly uses the `inference` parameter, passes it as a proper typed `np.bool_`, and includes a robust graph check that ensures:

* `Not` consumes `inference`,
* `Dropout` receives `Not`'s output as its third input,
* The dropout ratio is correctly set.

**Comment by enpasos** [comment]
🚨 I'm rewriting `jax2onnx` history on **Tue, 1 June, 20:00 CEST** to remove large files (`docs/onnx/`) — they’re moving to Hugging Face for easier hosting and faster cloning.
Please `git push --all && git push --tags` **before then** to avoid losing local work; afterward, re-clone or reset your local repo.

**Comment by enpasos** [comment]
@clementpoiret I just realized that I unintentionally merged a version of your `origin/equinox` branch from this morning into `main` — sorry about that.

I was struggling to get the `gpt` branch merged in, especially with the large `.onnx` files. Once we move those to Hugging Face, things should be much easier.

Two options — totally up to you:

1. I can revert the merge on `main` if you’d prefer to keep `equinox` separate for now, especially if you plan to PR it later in a more structured way.
2. If you were planning to merge into `main` soon anyway, we can leave it as-is, and you can open a pull request with the remaining parts when you're ready.

Let me know what works best for you!

**Comment by clementpoiret** [comment]
Nice :)
For the activation functions, I had in mind to add them too. 
For the merge into main, I was waiting to have all norm layers before making the first PR, then do a second one for jax.nn activation functions.

I can break them into smaller, more frequent ones if you think it's better

**Comment by enpasos** [comment]
🚨 All `*.onnx` files have now been completely removed from the Git history. These files are only available locally and on Hugging Face going forward. To be on the safe side, I recommend re-cloning your local repository — that’s what I did as well.

**Comment by enpasos** [comment]
**Congrats on contributing Equinox support — really great work! 🎉**

I have to admit there's one issue, and the blame is entirely on my side. I recently realized that our test system wasn't fully comparing JAX and ONNX outputs numerically, which allowed some bugs to slip through unnoticed under passing tests.

This has now been fixed on the `v0.7` branch, and as a result, previously hidden bugs — including a few in the Equinox components — are now being caught by the improved numeric checks.

If you're okay with it, I can take care of fixing them alongside the other issues that surfaced. But if you’d prefer to take a look yourself, feel free — just let me know how you’d like to handle it.

I'm planning to finish the remaining fixes by **tomorrow evening (11.07.2025)** before heading off for a one-week mountain bike tour 🚵‍♂️, so feel free to jump in or leave it with me — whichever works best for you!

**Comment by enpasos** [comment]
I'll begin by fixing the non-Equinox components first.
Unless I hear otherwise from you, I’ll also go ahead and fix the Equinox parts shortly after. Hope that’s okay — just wanted to give you a heads-up in case you were planning to jump in.

**Comment by clementpoiret** [comment]
Hey! Great catch and sorry for not having seen that before too, I usually just open the resulting graph and manually check its structure and attributes

I won't have the time until next week unfortunately, so if it fits your schedule, it would be awesome if you could fix them.

A bit early but have a great week, sounds like a nice program :)

**Comment by enpasos** [comment]
I’ve fixed the trivial `nn` components and temporarily commented out the test cases for the Equinox components with a `TODO` marker. That way, they won’t appear in the component table in the README or in the v0.7.1 release notes.

Once we’ve addressed the remaining issues, we can properly announce Equinox initial support.

... and as of v0.7.1, the test system now correctly catches numeric differences between ONNX functions and their JAX counterparts.

**Comment by enpasos** [comment]
I’ve fixed the remaining issues — they turned out to be relatively minor. Everything is now pushed to the `equinox` branch. Feel free to take a look, and if you’re happy with it (or after any tweaks you’d like to make), we can go ahead and release this as the first batch of supported Equinox components.

**Comment by clementpoiret** [comment]
Awesome, thanks a lot. This time it's my turn to take some days off. I won't be able to contribute until next week

**Comment by enpasos** [comment]
Equinox branch is yours — I’ll just make sure not to break any tests on it. Enjoy the days off!
